lib.coffee

```diff
@@ -308,8 +308,18 @@ class globals.Document
 
       fieldsWalker @meta.fields
 
-      referencesIncludeProjection = LocalCollection._compileProjection referencesInclude
       referencesExcludeProjection = LocalCollection._compileProjection referencesExclude
+      referencesIncludeModification = (document) ->
+        # We can't use _compileProjection here as this would cause top-level fields to
+        # be overwritten when updating subfields.
+        result = {}
+        for path, value of referencesInclude
+          continue if value isnt 1
+
+          value = _.reduce path.split('.'), ((doc, atom) -> doc[atom]), document
+          result[path] = value unless _.isUndefined value
```

Comment:
I don't think this `_.isUndefined value` is necessary, this are not fields from `observeChanges`? `undefined` will here be converted to `null` upon insertion. So if you would do normal insert with a document where field is `undefined`, would Meteor insert that field and make it `null`? If yes, then we should do it here as well.